commons-core/src/main/scala/com/avsystem/commons/rpc/OptionLike.scala

```diff
@@ -0,0 +1,43 @@
+package com.avsystem.commons
+package rpc
+
+sealed trait OptionLike[O] {
+  type Value
+  def none: O
+  def some(value: Value): O
+  def fold[B](opt: O, ifEmpty: => B)(f: Value => B): B
+}
+trait BaseOptionLike[O, A] extends OptionLike[O] {
```

Comment:
I use the base trait to do implicit search in macros. This way I avoid using wildcards and the implicit search is simpler. At the same time, I want to avoid the base trait to be extended directly because someone might forget about setting the type member to concrete type. For this reason, `OptionLike` is sealed while `BaseOptionLike` is not.